text/0000-unicode-string.md

```diff
@@ -0,0 +1,100 @@
+
+The following additional changes will be made to the String class:
+1. The size() function will return the number of unicode codepoints in the string. A new byte_size() function will return the number of bytes.
+1. The truncate() function will only support len parameter values less than the string size. The len parameter defines a number of codepoints. Add a resize() function to set the size of a string in bytes. This is needed for FFI calls where the _ptr is populated by the function, and the string length in bytes is provided by the function.
+1. The utf32() function will be removed. It is redundant, and returns pair that includes a byte count that is no longer needed.
+1. The insert_byte() function will be changed to insert_utf32()
+1. The values() function will return an iterator over the string codepoints. Same as runes(). A new bytes() function will return an iterator over string encoded as bytes. The bytes() function will take a StringEncoder parameter.
+1. A concat_bytes() function will be added to add a sequence of codepoints to the string from an iterator of bytes.
+1. Change 'fun val array: Array[U8] val' to 'fun array: this->Array[U8] box'. This is a superset of what we have now, and allows a readable "byte string" reference to a String ref, rather than requiring val.
+
+Add whatever methods we need to add to Array[U8] that are restricted by A:U8 (e.g. read_u8) to make Array[U8] have everything that a "byte string" class needs.
+
+Add traits StringEncoder and StringDecoder to the builtin package. Any function that produces a String from bytes, or produces bytes from a String must take a StringEncoder or StringDecoder as a type parameter as is appropriate. 
+```
+trait val StringEncoder
+  fun encode(codepoint: U32): (USize, U8, U8, U8, U8)
+  
+trait val StringDecoder
+  fun decode(bytes: U32): (U32, U8)
+```
+
+The ByteSeq type defined in std_stream.pony will be changed to remove String.
+```
+type ByteSeq is (Array[U8] val)
+```
+Many functions that accept ByteSeq as a parameter will be changed to accept (String | ByteSeq) as a parameter.
+
+A new StringIter interface will be added to std_stream.pony
+```
+interface val StringIter
+  """
+  An iterable collection of String box.
+  """
+  fun values(): Iterator[this->String box]
+```
+
+Change Reader in buffered/reader.pony to add functions to read a codepoint and to read a String of a given number of codepoints. Update function line() to accept a decoder, and to return a pair with the line string, and the number of bytes consumed.
+
+Change Writer in buffered/writer.pony to accept StringEncoder parameters in the write() and writev() functions. 
+
+Add a FileCharacters class in files/file_characters.pony that provides an iterator of characters in a file. The implementation will be similar to the FileLines class in files/file_lines.pony.
```

Comment:
Since the Unicode standard does not define what a character is, I think it's worth claryfing if it means codepoints, grapheme clusters or something else